File: 2018/04-sleeping-guards/go/star1.go
package main

import (
	"fmt"
	"io/ioutil"
	"sort"
	"strings"
)

func main() {
	bytes, _ := ioutil.ReadFile("input")
	lines := strings.Split(string(bytes), "\n")
	lines = lines[:len(lines)-1]
	sort.Strings(lines)

	var year, month, day, hour int // only used for correct parsing
	var guard, minute, sleepStart int
	sleepMinutes := make(map[int][]int)
	for _, line := range lines {
		fmt.Sscanf(line, "[%d-%d-%d %d:%d]", &year, &month, &day, &hour, &minute)
		action := line[19:]

		switch {
		case action[0:1] == "G":
			fmt.Sscanf(action, "Guard #%d begins shift", &guard)
		case action == "falls asleep":
			sleepStart = minute
		case action == "wakes up":
			for i := sleepStart; i < minute; i++ {
				sleepMinutes[guard] = append(sleepMinutes[guard], i)
			}
		}
	}

	var sleepiestGuard, sleepTime int
	for guard, minutes := range sleepMinutes {
		if len(minutes) > sleepTime {
			sleepiestGuard = guard
			sleepTime = len(minutes)
		}
	}

	minuteSlept := make(map[int]int)
	for _, minute := range sleepMinutes[sleepiestGuard] {
		minuteSlept[minute] += 1
	}

	var sleepiestMinute, max int
	for minute, times := range minuteSlept {
		if times > max {
			max = times
			sleepiestMinute = minute
		}
	}
	fmt.Println(sleepiestGuard * sleepiestMinute)
}

File: 2018/04-sleeping-guards/rust/star1.rs
use std::collections::HashMap;

const INPUT: &str = include_str!("../input");

fn parse(line: &str) -> (&str, u32) {
    let splitted: Vec<&str> = line.split(']').collect();
    let action = &splitted[1][1..];
    let minute: u32 = splitted[0][15..].parse().unwrap();
    return (&action, minute);
}

fn measure_sleep_times(
    sorted_lines: Vec<&str>,
) -> (HashMap<u32, u32>, HashMap<u32, HashMap<u32, u32>>) {
    let mut guard_sleep_time: HashMap<u32, u32> = HashMap::new();
    let mut guard_minute_asleep: HashMap<u32, HashMap<u32, u32>> = HashMap::new();
    let mut current_guard = 0;
    let mut sleep_start = 0;

    for line in sorted_lines {
        let (action, minute) = parse(line);
        if action.starts_with("Guard #") {
            current_guard = action.split(' ').nth(1).unwrap()[1..].parse().unwrap();
        } else if action == "falls asleep" {
            sleep_start = minute;
        } else if action == "wakes up" {
            let sleep_time = minute - sleep_start;
            *guard_sleep_time.entry(current_guard).or_default() += sleep_time;
            for min in sleep_start..minute {
                *guard_minute_asleep
                    .entry(current_guard)
                    .or_default()
                    .entry(min)
                    .or_default() += 1;
            }
        }
    }

    return (guard_sleep_time, guard_minute_asleep);
}

fn main() {
    let mut lines: Vec<&str> = INPUT.lines().collect();
    lines.sort();

    let (guard_sleep_time, guard_minute_asleep) = measure_sleep_times(lines);

    let mut sleepiest_guard: (&u32, &u32) = (&0, &0);
    for guard in guard_sleep_time.iter() {
        if guard.1 > sleepiest_guard.1 {
            sleepiest_guard = guard;
        }
    }
    let sleepiest_guard = sleepiest_guard.0;

    let mut sleepiest_minute: (&u32, &u32) = (&0, &0);
    for minute in guard_minute_asleep.get(sleepiest_guard).unwrap().iter() {
        if minute.1 > sleepiest_minute.1 {
            sleepiest_minute = minute;
        }
    }
    let sleepiest_minute = sleepiest_minute.0;

    println!("{:?}", sleepiest_guard * sleepiest_minute);
}
